mne/inverse_sparse/tests/test_mxne_optim.py

```diff
@@ -118,6 +120,51 @@ def test_tf_mxne():
     assert_array_equal(np.where(active_set_hat_tf)[0], active_set)
 
 
+def test_dgapl21l1():
+    """Test duality gap for L21 + L1 regularization."""
+    n_orient = 2
+    M, G, active_set = _generate_tf_data()
+    n_times = M.shape[1]
+    n_sources = G.shape[1]
+    tstep, wsize = 4, 32
+    n_steps = int(np.ceil(n_times / float(tstep)))
+    n_freqs = wsize // 2 + 1
+    n_coefs = n_steps * n_freqs
+    phi = _Phi(wsize, tstep, n_coefs)
+    phiT = _PhiT(tstep, n_freqs, n_steps, n_times)
+
+    for l1_ratio in [0.1, 0.3, 0.5, 0.8]:
+        alpha_max = norm_epsilon_inf(G, M, phi, l1_ratio, n_orient)
+        alpha_space = (1. - l1_ratio) * alpha_max
+        alpha_time = l1_ratio * alpha_max
+
+        Z = np.zeros([n_sources, n_coefs])
+        shape = (-1, n_steps, n_freqs)
+        # for alpha = alpha_max, Z = 0 is the solution so the dgap is 0
+        gap = dgap_l21l1(M, G, Z, np.ones(n_sources, dtype=bool),
+                         alpha_space, alpha_time, phi, phiT, shape, n_orient,
+                         -np.inf)[0]
+
+        assert_allclose(0., gap)
+        # check that solution for alpha smaller than alpha_max is non 0:
+        X_hat_tf, active_set_hat_tf, E, gap = tf_mixed_norm_solver(
+            M, G, alpha_space / 1.01, alpha_time / 1.01, maxit=200, tol=1e-8,
+            verbose=True, debias=False, n_orient=n_orient, tstep=tstep,
+            wsize=wsize, return_gap=True)
+        # allow possible small numerical errors (negative gap)
+        assert_array_less(-1e-10, gap)
+        assert_array_less(gap, 1e-8)
+        assert_array_less(1, len(active_set_hat_tf))
+
+        X_hat_tf, active_set_hat_tf, E, gap = tf_mixed_norm_solver(
+            M, G, alpha_space / 10, alpha_time / 10, maxit=200, tol=1e-8,
+            verbose=True, debias=False, n_orient=n_orient, tstep=tstep,
+            wsize=wsize, return_gap=True)
+        assert_array_less(-1e-10, gap)
+        assert_array_less(gap, 1e-8)
+        assert_array_less(1, len(active_set_hat_tf))
```

Comment:
how long does this take to run?

Reply:
220 seconds. But I believe testing l1_ratio = 0.8 is unrealistic in fact. When I wrote it I wanted to test something below and above 0.5.
If we want to speed up we could also use tol=1e-6.

Reply:
Removing 0.8 (and probably even 0.5) is fine for me. Instead we could add a ratio below 0.1, which is closer to what I typically used with TF-MxNE.

Reply:
I've put 0.05 and 0.1 only, and used higher alphas. Running time is now < 10s.